src/Microsoft.AspNetCore.SignalR.Redis/Internal/RedisProtocol.cs

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Microsoft.AspNetCore.SignalR.Internal;
+using Microsoft.AspNetCore.SignalR.Internal.Protocol;
+
+namespace Microsoft.AspNetCore.SignalR.Redis.Internal
+{
+    public static class RedisProtocol
+    {
+        private static readonly Encoding _utf8NoBom = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
+
+        // REVIEW: This protocol is simplistic and probably inefficient as is. Discuss.
+
+        // The Redis Protocol:
+        // * The message type is known in advance because messages are sent to different channels based on type
+        // * Invocations are sent to the All, Group, Connection and User channels
+        // * Group Commands are sent to the GroupManagement channel
+        // * Acks are sent to the Acknowledgement channel.
+        // * See the Write[type] methods for a description of the protocol for each in-depth.
+        // * The "Variable length integer" is the length-prefixing format used by BinaryReader/BinaryWriter:
+        //   * https://docs.microsoft.com/en-us/dotnet/api/system.io.binarywriter.write?view=netstandard-2.0
+        // * The "Length prefixed string" is the string format used by BinaryReader/BinaryWriter:
+        //   * A 7-bit variable length integer encodes the length in bytes, followed by the encoded string in UTF-8.
+
+        public static byte[] WriteInvocation(string methodName, object[] args, IReadOnlyList<IHubProtocol> protocols) =>
+            WriteInvocation(methodName, args, excludedIds: null, protocols);
+
+        public static byte[] WriteInvocation(string methodName, object[] args, IReadOnlyList<string> excludedIds, IReadOnlyList<IHubProtocol> protocols)
+        {
+            // Redis Invocation Format:
+            // * Variable length integer: Number of excluded Ids
+            // * For each excluded Id:
+            //   * Length prefixed string: ID
+            // * SerializedHubMessage encoded by the format described by that type.
+
+            using (var stream = new MemoryStream())
```

Comment:
Use pooled memory stream?

Reply:
Yeah, I was thinking of using something like the `LimitBlahBlahStream` (I forget the specific name and am lazy). Is that appropriate for this or should we just pool the `MemoryStream` itself?

Reply:
We could consider use the MemoryBufferWriter and wrapping that in a Stream. It'll basically replace the LimitPooledMemoryStream or whatever 😄

Reply:
Isn't that just adding another layer of abstraction?

Reply:
If we're going to do that, I'd prefer to just use `IBufferWriter<byte>` directly. This is a custom format and fairly simple, so it's not like we have to go through gymnastics to get to a Stream.

Reply:
Doesn't `BinaryWriter` and `BinaryReader` expect a `Stream`?

Reply:
Yeah, if we want to proceed with this as the protocol. We originally talked about using something like MsgPack or just our own custom protocol for this. I used `BinaryWriter` to get something that worked right away.